src/expr/proof.h

```diff
@@ -0,0 +1,178 @@
+ * B, that were already given proof steps.
+ *
+ * This class requires a ProofNodeManager object, which is a trusted way of
+ * allocating ProofNode pointers. This manager may have an underlying checker,
+ * although this is not required. It also (optionally) takes a context c.
+ * If no context is provided, then this proof is context-independent. This
+ * is implemented internally by using a dummy context that is never pushed
+ * or popped. The map from Nodes to ProofNodes is context-dependent and is
+ * backtracked when its context backtracks.
+ */
+class CDProof
+{
+  typedef context::CDHashMap<Node, std::shared_ptr<ProofNode>, NodeHashFunction>
+      NodeProofNodeMap;
+
+ public:
+  CDProof(ProofNodeManager* pnm, context::Context* c = nullptr);
+  ~CDProof();
+  /**
+   * Get proof for fact, or nullptr if it does not exist. Notice that this call
+   * does *not* clone the ProofNode object. Hence, the returned proof may
+   * be updated by further calls to this class. The caller should call
+   * ProofNode::clone if they want to own it.
+   */
+  std::shared_ptr<ProofNode> getProof(Node fact) const;
+  /** Add step
+   *
+   * @param expected The intended conclusion of this proof step. This must be
+   * non-null.
+   * @param id The identifier for the proof step.
+   * @param children The antecendant of the proof step. Each children[i] should
+   * be a fact previously added as a conclusion of a addStep call
+   * when ensureChildren is true.
+   * @param args The arguments of the proof step.
+   * @param ensureChildren Whether we wish to ensure steps have been added
+   * for all nodes in children
+   * @param forceOverwrite Whether we wish to overwrite if a step for expected
+   * was already provided (via a previous call to addStep)
+   * @return The true if indeed the proof step proves expected, or
+   * false otherwise. The latter can happen if the proof has a different (or
+   * invalid) conclusion, or if one of the children does not have a proof and
+   * ensureChildren is true.
+   *
+   * This method may create proofs justified by ASSUME of the facts in
+   * children that have not already been proven when ensureChildren is false.
+   * Notice that ensureChildren should be true if the proof is being
+   * constructed is a strictly eager fashion (bottom up from its leaves), while
+   * ensureChildren should be false if the steps are added lazily or out
+   * of order.
+   *
+   * This method only overwrites proofs for facts that were added as
+   * steps with id ASSUME when forceOverwrite is false, and otherwise always
+   * overwrites an existing step if one was provided when forceOverwrite is
+   * true.
+   */
+  bool addStep(Node expected,
+               PfRule id,
+               const std::vector<Node>& children,
+               const std::vector<Node>& args,
+               bool ensureChildren = false,
+               bool forceOverwrite = false);
+  /** Add proof
+   *
+   * @param fact The intended conclusion of the proof.
```

Comment:
I think we can do away with passing expected here. The proof node has a conclusion and the expected node is passed just to check that the conclusion of the proof is *already* equal to it.